### ISSUE TITLE
Problem with PATCH /vendors/{id}

### ISSUE THREAD

**Issue by liutingdu** [evercam/evercam-api#406]
I was trying to add a new MAC address '44:19:B6' to the existing vendor 'hikvision' from swagger, I see the doc says about MAC address is:

> Comma separated list of MAC's prefixes the vendor uses

So I thought I need to send all MAC addresses including the ones that already exists, then after sending the request, I get duplicate MACs : 

```
"known_macs": [
        "00:0C:43",
        "00:40:48",
        "8C:E7:48",
        "00:3E:0B",
        "44:19:B7",
        "00:0C:43",
        "00:40:48",
        "8C:E7:48",
        "00:3E:0B",
        "44:19:B7",
        "44:19:B6"
      ],
```

So the problems are:
- Is it possible to remove duplicate MAC addresses from vendor?
- How can I remove a MAC address from a vendor?
- I can easily PATCH vendor by ID from swagger, does this mean all users are able to do this? If it does, then it's dangerous because discovery is relying on those data.

@marcoherbst I am not sure who is the right person to answer question like this :)

**Comment by mosic** [comment]
Nice issue @liutingdu. What you tried is logical, I'd think the same. I just checked, we are only appending new mac adresses we get from PATCH vendors route, you can see it in [vendor_update.rb#L33](https://github.com/evercam/evercam-api/blob/3e190d4369e44dce6760088b6e95fab13dbcc041/lib/actors/vendor_update.rb#L33) file.

And to answer your question about the current state:

* It's not possible right now via the API.
* It's not possible right now via the API.
* Yes, you're right.

@azharmalik3 implemented this, so he may be the right person to improve it.

**Comment by marcoherbst** [comment]
Well done @mosic @liutingdu

The most urgent question is in respect to anyone being able to modify
vendor / model data. I guess we better limit this to admins. Pity. As
before, I think the way we store Vendor/Model info needs rethinking to make
it more in tune with Github pull requests.

As an aside, I think I see where the original idea for concatenating MACs
was coming from though. The common use case will be adding MACs. A bad
scenario is to allow easy overwriting of data. LD's situation is relatively
easy to solve with a DELETE and a POST (? is it?)....

Let's talk later about this.

M

On 18 August 2015 at 05:25, Miloš Mošić <notifications@github.com> wrote:

> Nice issue @liutingdu <https://github.com/liutingdu>. What you tried is
> logical, I'd think the same. I just checked, we are only appending new mac
> adresses we get from PATCH vendors route, you can see it in
> vendor_update.rb#L33
> <https://github.com/evercam/evercam-api/blob/3e190d4369e44dce6760088b6e95fab13dbcc041/lib/actors/vendor_update.rb#L33>
> file.
>
> And to answer your question about the current state:
>
>    - It's not possible right now via the API.
>    - It's not possible right now via the API.
>    - Yes, you're right.
>
> @azharmalik3 <https://github.com/azharmalik3> implemented this, so he may
> be the right person to improve it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-api/issues/406#issuecomment-132069290>
> .
>



-- 
Marco Herbst
*Evercam.io <http://www.evercam.io/>* - Cameras Unleashed
Mobile: +353 86 196 0451
Skype: marco.a.herbst

**Comment by marcoherbst** [comment]
Ok, after discussion, we're moving this functionality (and any other that modify Vendor/Models) to the admin area.

I'm going to suggest that @azharmalik3 takes this task.

Well done all.

**Comment by azharmalik3** [comment]
@marcoherbst @mosic @liutingdu Vendor/models add/modify functionality moved to admin area. Now you can edit any vendor/model from admin area. PATCH endpoints for vendor/model are still in API doc, will be removed soon.

**Comment by marcoherbst** [comment]
Excellent work, well done.
M

On 21 August 2015 at 12:51, Azhar Malik <notifications@github.com> wrote:

> Closed #406 <https://github.com/evercam/evercam-api/issues/406>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-api/issues/406#event-388714231>.
>



-- 
Marco Herbst
*Evercam.io <http://www.evercam.io/>* - Cameras Unleashed
Mobile: +353 86 196 0451
Skype: marco.a.herbst

**Comment by liutingdu** [comment]
@azharmalik3  Thanks a lot! 

It works for me now, I've fixed the MAC addresses that I added by mistake and I'll also add some new models.
